### DISCUSSION TITLE
Limit empty mileage

### DISCUSSION THREAD

**Discussion by JFDionne** [google/or-tools#3836] [Routing (and legacy CP) questions]
Hi, what will be the best way to limit the "empty milage" to solution?
Is there a setting for set the maximum milage without anything in the vehicle?

**Comment by jmarca** [comment]
That is not easy to do, but there are tricks. However, you need to provide more context. What does empty mean, and why would a vehicle be full or empty?You have to be careful. If you’re successful, you run the risk of just toting around a load for no reason just to minimize empty miles. JamesOn Jun 27, 2023, at 18:12, JFDionne ***@***.***> wrote:﻿
Hi, what will be the best way to limit the "empty milage" to solution?
Is there a setting for set the maximum milage without anything in the vehicle?

**Comment by Mizux** [comment]
AFAIK we don't have builtin support for it.

In C++ I would try:

Proposal 1: use `AddDimensionDependentDimension` on the CapacityDimension
i.e.
* `CapacityDimension.CumulVar == 0` -> return the distance_matrix[from][to].
* `CapacityDimension.CumulVar != 0` -> return 0.

BUT it will be pain full + if your capacity dimension is in range `[0, M]` while basically you have two cases:  `0` and `[1,M]` this will lead to precomputation of `N*N*M`values (with N being the number of nodes) while `N*N*2` whould have been sufficient.

Proposal 2: try to use a 0 ConstantDimension with SlackVar and add constraint on SlackVar value BUT it will be difficult to compute the constraints.
e.g. pseudo code
```
for all i in indices and not being a vehicle end:
 add constraint: if NextVar(i) == j and Capacitydim.CumulVar(j) != 0 then SlackVar(i) == distance(i, j)
```

**Comment by jmarca** [comment]
Consider the following case: two deliveries A and B in a straight line out from the depot. A is close to depot, B is far

For pure shortest path goals, there is no difference between deliver A, then B, or deliver B, then A. 

If you minimize empty distance, then the best choice is to deliver to B, then to A. 

Is that what you want, and why?

I solved a similar (but reversed) problem years ago by considering the distance to depot and ranking the delivery priorities. But that was a pickup and delivery problem with time windows, so doing so a priori was possible as I could exclude irrelevant destinations (non overlapping time windows, pickup points, etc).

**Comment by JFDionne** [comment]
Hi, I'll give you a simple example I found, not sure if it will work: (in python).


```python
from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp

from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp

def create_data_model():
    """Creation of modele with distance and kms running"""
    data = {}
    data['distance_matrix'] = [
        [0, 10, 15],   # Distance city 0 to city 1, 2 and 3
        [10, 0, 20],   # Distance city 1 to city 0, 2 and 3
        [15, 20, 0]    # Distance city 2 to city 0, 1 and 3
    ]
    data['empty_mileage'] = [5, 10, 8]  # KM when the car is empty
    return data

def main():
    data = create_data_model()

    manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']), 1, 0)
    routing = pywrapcp.RoutingModel(manager)

    # Add distance in a cost mode
    def distance_callback(from_index, to_index):
        # The price for the distance + KM running when it empty
        return data['distance_matrix'][manager.IndexToNode(from_index)][manager.IndexToNode(to_index)] + data['empty_mileage'][manager.IndexToNode(from_index)]

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

    # Add limitation KM constraint
    empty_mileage_limit = 12
    for vehicle_id in range(routing.vehicles()):
        routing.AddDimension(
            transit_callback_index,
            0,  # no slack
            empty_mileage_limit,  # limitation KM wen the car is empty
            False,  # start cumul to zero
            'EmptyMileage')
        empty_mileage_dimension = routing.GetDimensionOrDie('EmptyMileage')
        empty_mileage_dimension.CumulVar(routing.Start(vehicle_id)).SetRange(empty_mileage_limit, empty_mileage_limit)

    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)

    solution = routing.SolveWithParameters(search_parameters)

    if solution:
        print_solution(manager, routing, solution)

def print_solution(manager, routing, solution):
    """Show solution."""
    print('Distance totale : {} kilomètres'.format(solution.ObjectiveValue()))
    for vehicle_id in range(routing.vehicles()):
        index = routing.Start(vehicle_id)
        plan_output = 'Route for vehicle {} : '.format(vehicle_id)
        route_distance = 0
        while not routing.IsEnd(index):
            plan_output += ' {} ->'.format(manager.IndexToNode(index))
            previous_index = index
            index = solution.Value(routing.NextVar(index))
            route_distance += routing.GetArcCostForVehicle(previous_index, index, vehicle_id)
        plan_output += ' {}\n'.format(manager.IndexToNode(index))
        route_distance += routing.GetArcCostForVehicle(previous_index, index, vehicle_id)
        plan_output += 'Distance du véhicule {} : {} kilomètres\n'.format(vehicle_id, route_distance)
        print(plan_output)

# run main 
if __name__ == '__main__':
    main()
```
